pulser-simulation/pulser_simulation/simulation.py

```diff
@@ -620,6 +615,43 @@ def _update_noise(self) -> None:
             )
             self._doppler_detune = dict(zip(self._qid_index, detune))
 
+        idx_depolarizing = idx_dephasing = -1
+        for idx, collapse_op in enumerate(self._collapse_ops):
+            if "depolarizing" in self.config.noise and "X" in str(collapse_op):
+                idx_depolarizing = idx
+            if "dephasing" in self.config.noise and "Z" in str(collapse_op):
+                idx_dephasing = idx
+
+        if "depolarizing" in self.config.noise and idx_depolarizing != -1:
+            prob = self.config.depolarizing_prob / 4
+            n = self._size
+            if prob > 0.1 and n > 1:
+                warnings.warn(
+                    "The depolarizing model is a first-order approximation"
+                    f" in the depolarizing probability. p = {4*prob}"
+                    " is too large for realistic results.",
+                    stacklevel=2,
+                )
+            k = np.sqrt((prob) * (1 - 3 * prob) ** (n - 1))
+            for idx in range(idx_depolarizing, idx_depolarizing + 3):
+                self._collapse_ops[
+                    idx
+                ] *= k  # Update X, Y, Z collapse operators
+        if "dephasing" in self.config.noise and idx_dephasing != -1:
+            prob = self.config.dephasing_prob / 2
+            n = self._size
+            if prob > 0.1 and n > 1:
+                warnings.warn(
+                    "The dephasing model is a first-order approximation in the"
+                    f" dephasing probability. p = {2*prob} is too large for "
+                    "realistic results.",
+                    stacklevel=2,
+                )
+            k = np.sqrt(prob * (1 - prob) ** (n - 1))
+            self._collapse_ops[
+                idx_dephasing
+            ] *= k  # Update Z collapse operator
+
```

Comment:
I'm struggling to understand what you are trying here, but I can give you a few points as to why it's not the way to go:

1. `_update_noise()` is reserved for noises that rely on some form of random distrbution sampling (which is not the case for the collapse operators)
2. `self._collapse_ops` stores `qutip.Qobj`s, so I wouldn't rely on their `str()` like I see you did here.
3. I identify substantial code duplication here that I'm pretty sure can be avoided

I'm curious, did you test this? Did it work?

Reply:
Ah, I see, I interpreted the first review incorrectly, along with the requirements. It didn't work functionally, to do what we want (though I see now that I missed a save in the commit that made it fail entirely), but I started having doubts so I'm glad I checked in.

The latest comment certainly clarifies my confusion, so thanks! One thing of note, is that my initial approach added to the confusion. The first thing I did was remove the `NotImplementedError` blocks and apply the state elimination to the `density_matrix` (for which all tests pass), but that didn't seem to make sense in the context of the requirements.

Reply:
Ok, that makes sense! I was a bit puzzled so thanks for clarifying :)